### PR TITLE
dev-util/idea-ultimate (new ebuild)

### PR DESCRIPTION
**Pull request by switch87**
def-util/idea-ultimate: new ebuild (version 14.1.4 and 15_pre-release)

This does also fix bug [87419](https://bugs.gentoo.org/show_bug.cgi?id=87419)

Acked-by: Gert Pellin <gert@gepe-biljarts.be>
Reported-by: Gert Pellin <gert@gepe-biljarts.be>

### CODE REVIEW

**Comment by hasufell** [comment]
The merge commit is confusing, were you struggling with git? :)

Also check the commit message format:
https://wiki.gentoo.org/wiki/Gentoo_git_workflow#Example

File: dev-util/idea/idea-14.1.4.ebuild [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Copyright 1999-2015 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Header: $
```
Comment (hasufell) [inline dev-util/idea/idea-14.1.4.ebuild]: header is outdated, should be `# $Id$` now in git

File: dev-util/idea/idea-14.1.4.ebuild [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+EAPI=5
+
+DESCRIPTION="A complete toolset for web, mobile and enterprise development."
+HOMEPAGE="http://www.jetbrains.com/idea"
+SRC_URI="https://d1opms6zj7jotq.cloudfront.net/idea/ideaIU-${PV}.tar.gz"
```
Comment (hasufell) [inline dev-util/idea/idea-14.1.4.ebuild]: is that a reliable SRC_URI?

File: dev-util/idea/idea-15_pre.ebuild [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Copyright 1999-2015 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Header: $
```
Comment (hasufell) [inline dev-util/idea/idea-15_pre.ebuild]: same as above

File: dev-util/idea/idea-15_pre.ebuild [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+src_install() {
+	local dir="/opt/${PN}"
+
+	insinto "${dir}"
+	doins -r *
+	fperms 755 ${dir}/bin/{idea.sh,fsnotifier{,64}}
+
+	# Create /usr/bin/intelliJ
+	dodir /usr/bin/
+	cat <<-EOF >"${ED}"usr/bin/${PN}
```
Comment (hasufell) [inline dev-util/idea/idea-15_pre.ebuild]: I don't think `${ED}` is guaranteed to end with a `/`. Afaik, the best workaround is to do `"${ED%/}"/usr/bin/${PN}`

Comment (hasufell) [inline dev-util/idea/idea-15_pre.ebuild]: Also, it should be
```
cat <<-EOF > "${ED%/}"/usr/bin/${PN} || die "cat failed!"
```
for proper error handling

File: dev-util/idea/idea-14.1.4.ebuild [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+src_install() {
+	local dir="/opt/${PN}"
+
+	insinto "${dir}"
+	doins -r *
+	fperms 755 ${dir}/bin/{idea.sh,fsnotifier{,64}}
+
+	# Create /usr/bin/intelliJ
+	dodir /usr/bin/
+	cat <<-EOF >"${ED}"usr/bin/${PN}
```
Comment (hasufell) [inline dev-util/idea/idea-14.1.4.ebuild]: same as below

**Comment by hasufell** [comment]
Also, you should reference the bug [87419](https://bugs.gentoo.org/show_bug.cgi?id=87419)

**Comment by switch87** [comment]
sorry for the confusing commits, since your last view I changed the things you asked me to and I added the licenses.

**Comment by hasufell** [comment]
@ulm how do the licenses look?

**Comment by mrueg** [comment]
@switch87  Please use idea-community and idea-professional as PN if you want to move it to the tree. Might be worth to have a look at the ebuild available in my overlay: https://github.com/mrueg/mrueg-overlay/blob/master/dev-util/idea-community/idea-community-14.1.4.141.1532.4.ebuild
@johu might be interested in this, too.

**Comment by monsieurp** [comment]
We might opt for a source-based ebuild after all. @chewi successfully compiled IDEA using ant, yet without removing the content of the lib/ directory filled with a zillion jars, see https://github.com/JetBrains/intellij-community/tree/master/lib.
@switch87 you don't need to upload your ebuilds both here and on bugzilla ;) Give us a shout on bugzilla and create a PR here. That's fine.

**Comment by hasufell** [comment]
> We might opt for a source-based ebuild after all.

would it make sense to suffix the ebuilds with `-bin` then?

**Comment by chewi** [comment]
So I downloaded a master tarball from GitHub and built IDEA from source with just a mere invocation of ant. I only tried it briefly but the result appeared to work just fine.

The only thing it downloaded during the build was the Kotlin plugin and that could easily be fixed. There are tons of bundled jars in lib but even building from source without unbundling would be a good start. Having said that, a lot of these are things we already have packaged.

It doesn't look like our usual build.xml rewriting trick would work as it seems to source these jars using a collection of strange XML files under .idea/libraries. java-pkg_jar-from would probably be best here.

It's not clear how the libraries get sourced at runtime. It doesn't add the whole of lib to the classpath in the launcher script. Perhaps IDEA to hardcoded to look there. Would adding the unbundled jars to the initial classpath work or would it create problems like we've seen with Tomcat? Don't know. This might need some experimentation.

So to sum up, I won't accept a -bin version of this package given how easy it was to build from source. Sorry.

**Comment by switch87** [comment]
@chewi this is the professional version, only the community version can be build from source.

**Comment by chewi** [comment]
@switch87, sorry, kinda jumped in straight from Bugzilla and missed that point. You could still unbundle the jars but I'd be happy enough to only do that in the community version.

**Comment by hasufell** [comment]
We have to work a little bit on the commits. The licenses should be a separate commit IMO, and the commit message should look like:
```
licenses: add IDEA licenses for def-util/idea-professional
```

Then commit the ebuilds with a commit message of the form:
```
def-util/idea-professional: initial import wrt bug 87419
```

So we have only _two_ commits, no merge commits.

**Comment by mrueg** [comment]
Should be probably dev-util instead of def-util.

**Comment by switch87** [comment]
@hasufell stupid question, but how do I best do this? do I have to make desperate forks for the 2 commits? apparently everything I change in my fork adds to this commit.

@mrueg indeed it should ;)

**Comment by hasufell** [comment]
> stupid question, but how do I best do this? do I have to make desperate forks for the 2 commits? apparently everything I change in my fork adds to this commit.

make a __mixed reset__ (e.g. via the graphical `gitk` tool): right click on the commit you want to reset to (the one before you started your work), then click `Reset master branch to here`, select `mixed` and re-stage and re-commit (e.g. via the graphical tool `git gui`). That way you see exactly what is stages/commited and the history.

**Comment by hasufell** [comment]
you are almost there, just need to remove the `IntelliJ IDEA` since @chewi wants to do a source-based ebuild for that

**Comment by switch87** [comment]
At your service ;) without graphical tools

**Comment by hasufell** [comment]
we also need a `metadata.xml`, you can just create one, add it and then do `git commit --amend`

**Comment by switch87** [comment]
I added metadata.xml

File: dev-util/idea-professional/metadata.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
```
Comment (hasufell) [inline dev-util/idea-professional/metadata.xml]: add something like
```xml
	<herd>proxy-maintainers</herd>
	<maintainer>
		<email>your@email.com</email>
		<name>Your Name</name>
		<description>Proxy maintainer. Assign bugs to him.</description>
	</maintainer>
```

**Comment by sonOfRa** [comment]
Upstream calls the "professional" edition "ultimate". Maybe that should be reflected in the ebuild name?

**Comment by switch87** [comment]
@sonOfRa  you are correct, @mrueg, @hasufell  should I change it to ultimate?

**Comment by hasufell** [comment]
probably makes sense

**Comment by hasufell** [comment]
I think the last commit should be merged into the previous one (so we only have one ebuild commit).

**Comment by switch87** [comment]
Ok, now there are only the two commits that you asked for @hasufell

**Comment by hasufell** [comment]
I will merge this if there are no further objections.

**Comment by chewi** [comment]
Hold up, I've been told I have to review this. Won't be long.

**Comment by chewi** [comment]
Okay, it does fail repoman because IDEA_OpenSource LICENSE doesn't match the IDEA_Opensource filename case-wise.

@mrueg pointed that out DESCRIPTION shouldn't have a trailing dot!

Other than that, the only concern I have is that putting `>=virtual/jdk-1.7` gives you no guarantee that Java 6 won't be the selected VM when you try to run it. There's not a lot you can do about that short of making it use Gentoo's launcher scripts but that's not trivial to do here. We'll be kicking Java 6 out of the tree real soon now™ so I wouldn't worry about it.

**Comment by hasufell** [comment]
> it does fail repoman because

we really should have CI repoman hooks via travis or whatever, even if it's just for this repo

**Comment by switch87** [comment]
Ok, changed IDEA_Opensource to IDEA_OpenSource

**Comment by chewi** [comment]
Don't forget about the trailing dot. :wink: In case you think I'm being anal, I think some other dev did go over the whole tree just to remove them.

**Comment by switch87** [comment]
Ok, I deleted the dot in both ebuild files :sweat_smile:

**Comment by hasufell** [comment]
@chewi was that a +1?

**Comment by chewi** [comment]
Aye. :+1:

**Comment by hasufell** [comment]
@switch87 can you rebase your branch against the latest gentoo master?

E.g.
```
git pull --rebase=preserve git@github.com:gentoo/gentoo.git master
git push -f origin master
```

**Comment by switch87** [comment]
ok, I did the rebase

**Comment by switch87** [comment]
OK, my first ebuild merged, I'll try to do it in less steps next time :grin:  thanks for the help and patience :+1:
